File: technology-lab/java-showcase/src/test/java/no/ab/showcase/EventSummaryTest.java
package no.ab.showcase;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class EventSummaryTest {

    @Test
    void countsEventsByTypeAndTotalsBytes() {
        List<WorkEvent> events = List.of(
                new WorkEvent("e-1", WorkEvent.Type.CREATED, 512),
                new WorkEvent("e-2", WorkEvent.Type.UPDATED, 128),
                new WorkEvent("e-3", WorkEvent.Type.UPDATED, 256),
                new WorkEvent("e-4", WorkEvent.Type.DELETED, 0));

        EventSummary summary = EventSummary.of(events);

        assertEquals(4, summary.total());
        assertEquals(896L, summary.totalBytes());
        assertEquals(Map.of(
                WorkEvent.Type.CREATED, 1L,
                WorkEvent.Type.UPDATED, 2L,
                WorkEvent.Type.DELETED, 1L), summary.byType());
    }

    @Test
    void emptyInputGivesZeroCountsForEveryType() {
        EventSummary summary = EventSummary.of(List.of());

        assertEquals(0, summary.total());
        assertEquals(0L, summary.totalBytes());
        assertEquals(3, summary.byType().size());
    }

    @Test
    void rejectsInvalidEvents() {
        assertThrows(IllegalArgumentException.class, () -> new WorkEvent(" ", WorkEvent.Type.CREATED, 1));
        assertThrows(IllegalArgumentException.class, () -> new WorkEvent("e", null, 1));
        assertThrows(IllegalArgumentException.class, () -> new WorkEvent("e", WorkEvent.Type.CREATED, -1));
    }
}

File: technology-lab/kotlin-showcase/src/test/kotlin/no/ab/showcase/EventSummaryTest.kt
package no.ab.showcase

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class EventSummaryTest {
    @Test
    fun countsEventsByTypeAndTotalsBytes() {
        val events = listOf(
            WorkEvent("e-1", EventType.CREATED, 512),
            WorkEvent("e-2", EventType.UPDATED, 128),
            WorkEvent("e-3", EventType.UPDATED, 256),
            WorkEvent("e-4", EventType.DELETED, 0),
        )

        val summary = EventSummary.of(events)

        assertEquals(4, summary.total)
        assertEquals(896L, summary.totalBytes)
        assertEquals(mapOf(EventType.CREATED to 1, EventType.UPDATED to 2, EventType.DELETED to 1), summary.byType)
    }

    @Test
    fun emptyInputGivesZeroForEveryType() {
        val summary = EventSummary.of(emptyList())

        assertEquals(0, summary.total)
        assertEquals(EventType.entries.associateWith { 0 }, summary.byType)
    }

    @Test
    fun rejectsInvalidEvents() {
        assertFailsWith<IllegalArgumentException> { WorkEvent(" ", EventType.CREATED, 1) }
        assertFailsWith<IllegalArgumentException> { WorkEvent("e", EventType.CREATED, -1) }
    }
}
